source/common/event/scaled_range_timer.cc

```diff
@@ -0,0 +1,203 @@
+        : duration(duration), bucket_position(std::move(bucket_position)) {}
+    const MonotonicTime::duration duration;
+    const ScaledRangeTimerManager::BucketEnabledList::iterator bucket_position;
+  };
+
+  void onPendingTimerComplete() {
+    ENVOY_LOG_MISC(info, "pending complete for {}", static_cast<void*>(this));
+    ASSERT(absl::holds_alternative<Pending>(state_));
+    Pending& pending = absl::get<Pending>(state_);
+
+    const MonotonicTime now = manager_.dispatcher_.timeSource().monotonicTime();
+    auto it = manager_.add(*this, pending.latest_trigger - now);
+    state_.emplace<Active>(pending.latest_trigger - now, it);
+  }
+
+  ScaledRangeTimerManager& manager_;
+  const TimerCb callback_;
+  const TimerPtr pending_timer_;
+
+  absl::variant<Inactive, Pending, Active> state_;
+  const ScopeTrackedObject* scope_;
+};
+
+ScaledRangeTimerManager::ScaledRangeTimerManager(Dispatcher& dispatcher, float scale_factor,
+                                                 std::chrono::milliseconds minimum_duration)
+    : dispatcher_(dispatcher), minimum_duration_(minimum_duration), scale_factor_(scale_factor) {}
+
+RangeTimerPtr ScaledRangeTimerManager::createTimer(TimerCb callback) {
+  return std::make_unique<ScaledRangeTimer>(callback, *this);
+}
+
+void ScaledRangeTimerManager::setScaleFactor(float scale_factor) {
+  scale_factor_ = DurationScaleFactor(scale_factor);
+  for (auto& bucket : buckets_) {
+    if (!bucket.scaled_timers.empty()) {
+      bucket.updateTimer(*this);
+    }
+  }
+}
+
+MonotonicTime::duration
+ScaledRangeTimerManager::getBucketedDuration(MonotonicTime::duration max_duration) {
+  return MonotonicTime::duration(static_cast<MonotonicTime::duration::rep>(
+      std::pow(kBucketScaleFactor,
+               std::floor(std::log(max_duration.count()) / std::log(kBucketScaleFactor)))));
+}
+
+void ScaledRangeTimerManager::Bucket::updateTimer(ScaledRangeTimerManager& manager) {
+  if (scaled_timers.empty()) {
+    if (timer->enabled()) {
+      timer->disableTimer();
+    }
+  } else {
+    auto& entry = scaled_timers.front();
+    timer->enableTimer(std::chrono::duration_cast<std::chrono::milliseconds>(
+        manager.scale_factor_.value() *
+        (entry.latest_trigger_time - manager.dispatcher_.timeSource().monotonicTime())));
+  }
+}
+
+ScaledRangeTimerManager::DurationScaleFactor::DurationScaleFactor(float value)
+    : value_(value < 0 ? 0 : value < 1 ? value : 1) {}
+
+float ScaledRangeTimerManager::DurationScaleFactor::value() const { return value_; }
+
+ScaledRangeTimerManager::BucketEnabledList::iterator
+ScaledRangeTimerManager::add(ScaledRangeTimer& timer, MonotonicTime::duration max_duration) {
+  max_duration = getBucketedDuration(max_duration);
+  Bucket& bucket = getBucket(max_duration);
+
+  bucket.scaled_timers.emplace_back(timer, max_duration + dispatcher_.timeSource().monotonicTime());
+  bucket.updateTimer(*this);
```

Comment:
Calling updateTimer on every add will result in the bucket timers potentially never triggering.  After https://github.com/envoyproxy/envoy/pull/11823, a timer will never trigger in the same iteration as it is registered, even when the registration time delta is 0.  This re-registration results in the ball getting kicked down the road every time a range timer is added.

Reply:
I modified updateTimer to be smarter about when it modifies a timer. I think that should fix this bug and let us call updateTimer unconditionally.